src/Pool.sol

```diff
@@ -453,14 +453,113 @@ abstract contract Pool {
+        uint128 currAmtPerSec = _totalAmtPerSec(currReceivers);
+        uint64 currEndTime = _sendingEndTime(sender, currAmtPerSec);
+        withdrawn = _updateSenderBalance(sender, topUpAmt, withdrawAmt, currAmtPerSec, currEndTime);
+        sender.startTime = _currTimestamp();
+        uint64 newEndTime = _sendingEndTime(sender, newAmtPerSec);
+        updates = _updateStreams(currReceivers, newReceivers, currEndTime, newEndTime);
+    }
+
+    /// @notice Updates sender's `startBalance`.
+    /// @param sender The updated sender.
+    /// @param topUpAmt The topped up amount.
+    /// @param withdrawAmt The amount to be withdrawn, must not be higher than available funds.
+    /// Can be `WITHDRAW_ALL` to withdraw everything.
+    /// @param amtPerSec The sender's total amount per second
+    /// @param endTime Time when sending was supposed to stop.
+    /// @return withdrawn The withdrawn amount which should be sent to the user.
+    /// Equal to `withdrawAmt` unless `WITHDRAW_ALL` is used.
+    function _updateSenderBalance(
+        Sender storage sender,
+        uint128 topUpAmt,
+        uint128 withdrawAmt,
+        uint128 amtPerSec,
+        uint64 endTime
+    ) internal returns (uint128 withdrawn) {
+        if (endTime > _currTimestamp()) endTime = _currTimestamp();
+        sender.startBalance -= (endTime - sender.startTime) * amtPerSec;
+        sender.startBalance += topUpAmt;
+        if (withdrawAmt == WITHDRAW_ALL) withdrawAmt = sender.startBalance;
+        require(withdrawAmt <= sender.startBalance, "Not enough funds in the sender account");
+        sender.startBalance -= withdrawAmt;
+        return withdrawAmt;
+    }
+
+    /// @notice Updates streams in the receivers' `amtDeltas`.
+    /// @param currReceivers The list of the user's receivers which is currently in use.
+    /// If this function is called for the first time for the user, should be an empty array.
+    /// @param newReceivers The new list of the user's receivers.
+    /// @param currEndTime Time when sending using `currReceivers` was supposed to stop.
+    /// @param newEndTime Time when sending using `newReceivers` will be supposed to stop.
+    /// @return updates The list of stream updates to log
+    function _updateStreams(
+        Receiver[] calldata currReceivers,
+        Receiver[] calldata newReceivers,
+        uint64 currEndTime,
+        uint64 newEndTime
+    ) internal returns (StreamUpdates memory updates) {
+        // Skip iterating over `currReceivers` if funding has run out
+        uint256 currIdx = currEndTime > _currTimestamp() ? 0 : currReceivers.length;
+        // Skip iterating over `newReceivers` if no new funding is started
+        uint256 newIdx = newEndTime > _currTimestamp() ? 0 : newReceivers.length;
+        uint256 updatesMax = currReceivers.length + newReceivers.length - currIdx - newIdx;
+        updates = StreamUpdates({length: 0, updates: new StreamUpdate[](updatesMax)});
+        while (true) {
```

Comment:
This can be a bit simplified. Tests are passing.
```solidity
        while (currIdx < currReceivers.length || newIdx < newReceivers.length) {
            // Each iteration gets the next stream update and applies it on the receiver.
            // A stream update is composed of two receiver configurations, one from each list,
            // or from a single configuration if the receiver is only on one of the lists.
            bool pickCurr = currIdx < currReceivers.length;
            bool pickNew = newIdx < newReceivers.length;
            if (pickCurr && pickNew) {
                pickCurr = currReceivers[currIdx].receiver <= newReceivers[newIdx].receiver;
                pickNew = newReceivers[newIdx].receiver <= currReceivers[currIdx].receiver;
            }
            address receiver = address(0);
            int128 currAmt = 0;
            int128 newAmt = 0;
            if (pickCurr) {
                receiver = currReceivers[currIdx].receiver;
                currAmt = int128(currReceivers[currIdx].amtPerSec);
                _setDelta(receiver, currEndTime, currAmt);
                currIdx++;
            }
            if (pickNew) {
                receiver = newReceivers[newIdx].receiver;
                newAmt = int128(newReceivers[newIdx].amtPerSec);
                _setDelta(receiver, newEndTime, -newAmt);
                newIdx++;
            }
            _setDelta(receiver, _currTimestamp(), newAmt - currAmt);
            updates.updates[updates.length++] = StreamUpdate({
            receiver: receiver,
            amtPerSec: uint128(newAmt),
            endTime: newAmt == 0 ? _currTimestamp() : newEndTime
            });
            // The receiver was never used, initialize it.
            if (!pickCurr && receiverStates[receiver].nextCollectedCycle == 0) {
                receiverStates[receiver].nextCollectedCycle = _currTimestamp() / cycleSecs + 1;
            }
        }
```

Reply:
I've made a diff and the only change I can see here is that we repeatedly check relationship between the index and the list length instead of doing it once and reusing the result, is it simpler?

Reply:
I think it is cleaner to remove the `if else`  and move the condition when the loops stop into the `while(...)`

Reply:
Maybe we can also add the cases to the comments. From my notes: 

```javascript
// cases: 
// case 1 only currReceiver left: pick only currReceiver
// case 2 only nextReceiver left: pick only nextReceiver
// case 3 currReceiver and nextReceiver are left and both are equal: pickCurr and pickNext
// case 4 currReceiver and nextReceiver are left and both are not equal: pick smaller one
```

Reply:
> I think it is cleaner to remove the `if else`  and move the condition when the loops stop into the `while(...)`

The nice property of how it's expressed now is that it's clearly exhaustive, it's easy to see what should happen in case there's nothing to pick. If we remove `if else`, the algorithm is still exhaustive, but only by coincidence, which IMO is both harder to follow and easier to introduce bugs. I don't have very a strong opinion on that, if you're certain that your proposition will make ours and newcomers' lives easier, let's fix that.

Reply:
It is already highly optimised the trade-off is it is a bit hard to understand. 
Especially to see all the cases when `pickCurr` and `pickNew` are happening.

I think some more comments would be nice.   

I leave the decision for the `while` up to you. In the end it is taste

Reply:
Added some comments. The problem with this algorithm is that there are really not that many steps in it, but it's not trivial. It connects the `amtDelta`s updates, which is a tough concept on its own, with iteration matching pairs from two sorted lists, which is a rather uncommon way of consuming lists.